### PR TITLE
Add EdDSA signatures support.

### PR DESCRIPTION
**Pull request by plietar**
This adds support EdDSA signatures, when using OpenSSL.

Due to the way EdDSA works, an auxiliary buffer is need to serialize the Sig_Structure.
A pair of new functions, `t_cose_sign1_verify_set_sigstruct_buffer`, are used to provide
this buffer from the application to t_cose.

A "known good" signature, generated by pycose, is added to the test
suite. Additionally, the usual sign/verify roundtrip test is added.

### CODE REVIEW

**Comment by plietar** [comment]
@laurencelundblade I've finished adding the signing side of it. Please let me know what you think.

**Comment by laurencelundblade** [comment]
Paul, thanks for that!  I glad for all the algorithms support as it is helping me think about the t_cose 2.0 design. I will be merging all this stuff to the dev branch at some point.

It's going to take me a while to look at this because of other stuff I have going on.

**Comment by laurencelundblade** [comment]
Haven't looked through it all yet. I can compile it and run the tests.

There should be a DISABLE_EDDSA because lots of people won't want it and small code size is important.

As far as I know eddsa is the only signing algorithm that needs the auxiliary buffer so I think DISABLE_EDDSA should disable it too.  If it turns out that there is another algorithm that needs it, then we can make it so that it disables only if both those algorithms are disabled.

Still working on reviewing the whole PR in detail...

Thanks for your work here!

**Comment by plietar** [comment]
I didn't want the public API to change depending on whether EdDSA is available or not, which is why I didn't put the `set_auxiliary_buffer` behind an `ifndef`.

I did consider making the `set_auxiliary_buffer` function a no-op if EdDSA is not available and remove the pointer field from the signing and verification contexts. The function would get removed at link-time if it isn't used. This issue with this is that it breaks applications that do call the function, and expect the buffer size to be adjusted:

```c
q_useful_buf buffer = { NULL, INT32_MAX };
t_cose_sign_set_auxiliary_buffer(ctx, &buffer);
t_cose_sign1_sign(....); // Just measure the size
buffer.ptr = malloc(buffer.size);
t_cose_sign1_sign(....); // Do the actual signing
```

If `set_auxiliary_buffer` is a no-op, the application will try to allocate an `INT32_MAX` sized buffer. Another option would be, instead of a no-op, the `set_auxiliary_buffer` immediately sets the length to 0. But then when that length gets set depends on whether `DISABLE_EDDSA` is set, in pretty subtle ways.

-----

Honestly I'm not that happy with the `set_auxiliary_buffer` API, especially the way the signing operation has a non-obvious side effect. I'll try to think of something better.

**Comment by plietar** [comment]
I've made the mentioned API change, and I think it is a lot nicer to use than the previous one was. It also allows signing / verifying two different messages with the same context and auxiliary buffer, which the previous API did not (you'd have had to re-configure the auxiliary buffer between each operation).

File: examples/t_cose_basic_example_ossl.c [hunk 1]
```diff
@@ -528,17 +538,195 @@ int two_step_sign_example(void)
...
+    return_value = t_cose_sign1_sign(&sign_ctx,
+                                      constructed_payload,
+                                      (struct q_useful_buf){ NULL, SIZE_MAX },
+                                     &signed_cose);
+    printf("Computed signing size %d (%s)\n", return_value, return_value ? "fail" : "success");
+    if(return_value) {
+        goto Done;
+    }
+    printf("Output buffer size = %d bytes\n", signed_cose.len);
+    printf("Auxiliary buffer size = %d bytes\n", t_cose_sign1_sign_auxiliary_buffer_size(&sign_ctx));
```
Comment (laurencelundblade) [inline examples/t_cose_basic_example_ossl.c:615]: %zu to print size_t  Here and a few other places.

File: src/t_cose_util.h [hunk 2]
```diff
@@ -105,6 +126,33 @@ enum t_cose_err_t create_tbs_hash(int32_t                     cose_algorithm_id,
...
+ * Serialize the to-be-signed (TBS) bytes for COSE.
+ *
+ * \param[in] protected_parameters  Full, CBOR encoded, protected parameters.
+ * \param[in] aad                   Additional Authenitcated Data to be
+ *                                  included in TBS.
+ * \param[in] payload               The CBOR-encoded payload.
+ * \param[in] buffer_for_hash       Pointer and length of buffer into which
+ *                                  the resulting TBS bytes is put.
+ * \param[out] hash                 Pointer and length of the
+ *                                  resulting TBS bytes.
```
Comment (laurencelundblade) [inline src/t_cose_util.h:139]: buffer_for_hash and hash do not correspond to the actual variable names.

File: src/t_cose_sign1_sign.c [hunk 3]
```diff
@@ -315,6 +227,211 @@ t_cose_sign1_encode_parameters_internal(struct t_cose_sign1_sign_ctx *me,
...
+#endif /* T_COSE_DISABLE_EDDSA */
+
+/**
+ * \brief Compute a signature for a COSE_Sign1 message, following the
+ * general procedure which works for most algorithms.
+ *
+ * \param[in] me         The t_cose signing context.
+ * \param[in] aad        The Additional Authenticated Data or \c NULL_Q_USEFUL_BUF_C.
+ * \param[in] payload    Pointer and length of payload to sign.
+ * \param[in] out_buf    Pointer and length of buffer to output to.
```
Comment (laurencelundblade) [inline src/t_cose_sign1_sign.c:377]: should be buffer_for_signature

File: src/t_cose_sign1_verify.c [hunk 4]
```diff
@@ -308,51 +461,32 @@ t_cose_sign1_verify_internal(struct t_cose_sign1_verify_ctx *me,
...
 Done:
-    if(returned_parameters != NULL) {
-        *returned_parameters = parameters;
+    if (return_value == T_COSE_SUCCESS)
+    {
+        if(returned_parameters != NULL) {
+            *returned_parameters = parameters;
+        }
+        if(!is_dc && payload != NULL) {
+            *payload = signed_payload;
```
Comment (laurencelundblade) [inline src/t_cose_sign1_verify.c:488]: signed_payload is uninitialized here when header_decode fails above

File: test/t_cose_sign_verify_test.c [hunk 5]
```diff
@@ -129,6 +132,14 @@ int_fast32_t sign_verify_basic_test()
     }
 #endif
 
+#if !defined(T_COSE_DISABLE_EDDSA) && defined(T_COSE_CRYPTO_LIB_OPENSSL)
+    /** Only OpenSSL supports EDDSA. */
+    return_value  = sign_verify_basic_test_alg(T_COSE_ALGORITHM_EDDSA);
+    if(return_value) {
```
Comment (laurencelundblade) [inline test/t_cose_sign_verify_test.c:138]: I'd prefer the test run even if its PSA crypto or EDDSA is disabled. The expected results should be the unknown algorithm error.

Someday there will be measured code coverage (I hope) and these tests are needed to have the complete coverage.

File: test/t_cose_sign_verify_test.c [hunk 6]
```diff
@@ -8,6 +8,7 @@
  * See BSD-3-Clause license in README.md
  */
```
Comment (laurencelundblade) [inline test/t_cose_sign_verify_test.c:10]: There should be a test that tries to use Eddsa without setting the aux buffer to see that the right error is returned.

**Review by laurencelundblade — CHANGES_REQUESTED** [review]
This looks really good. I like the new design for aux buffer and moving short circuit sigs into its own source file is good. Thanks!

There's a few nits here to fix. It's good to have them fixed since this is going right into the commercial ready branch.

File: test/t_cose_sign_verify_test.c (hunk 5 above)
Comment (plietar) [reply test/t_cose_sign_verify_test.c]: Your comment has made me realize that `T_COSE_CRYPTO_LIB_OPENSSL` is not actually a preprocessor define but an enum value, which meant this test was never running.

There was actually no way to find out whether EDDSA is supported, so I picked up some of your changes from the dev branch to add the `t_cose_is_algorithm_supported`, and changed all the tests to use that. Let me know what you think of it.

I've added a separate test that checks for `ERR_UNSUPPORTED_SIGNING_ALG` instead.

File: test/t_cose_sign_verify_test.c (hunk 6 above)
Comment (plietar) [reply test/t_cose_sign_verify_test.c:10]: Done. Added both no aux buffer and also a buffer that is too small.

File: src/t_cose_sign1_sign.c (hunk 3 above)
Comment (plietar) [reply src/t_cose_sign1_sign.c]: Done.

File: src/t_cose_sign1_verify.c (hunk 4 above)
Comment (plietar) [reply src/t_cose_sign1_verify.c:492]: This line only executes if return_value == T_COSE_SUCCESS, so everything should be initialized correctly.

File: examples/t_cose_basic_example_ossl.c (hunk 1 above)
Comment (plietar) [reply examples/t_cose_basic_example_ossl.c]: Done.

File: src/t_cose_util.h (hunk 2 above)
Comment (plietar) [reply src/t_cose_util.h:139]: Done.

File: inc/t_cose/t_cose_sign1_verify.h [hunk 7]
```diff
@@ -264,6 +280,54 @@ t_cose_sign1_set_verification_key(struct t_cose_sign1_verify_ctx *context,
...
+ *
+ * This function can be called after \ref t_cose_sign1_verify (or
+ * equivalent) was called. If the context was initialized with the
+ * DECODE_ONLY flag, it returns the number of bytes that would have
+ * been used by the signing operation. This allows the caller to
+ * allocate an appropriately sized buffer before performing the
+ * actual verification.
+ *
+ * This function returns if the signature algorithm used does not need
+ * an auxiliary buffer.
```
[the remark this answers is not in the document]
Comment (plietar) [reply inc/t_cose/t_cose_sign1_verify.h]: Done.

**Comment by laurencelundblade** [comment]
There are some compiler errors still with the max warnings that tdv/b.sh uses.

I haven't found a way to use Q_USEFUL_BUF_FROM_BYTE_ARRAY_LITERAL()/sizeof() in static const data initializations. I haven't researched this thoroughly as to whether it is allowed by C99 or not, but I suspect it is not allowed. In the implementation of test cases it's OK if it is done at run time and thus has slightly larger object code.

**Comment by plietar** [comment]
`sizeof` seems to be fine in this context, it's the cast to `q_usebul_buf_c` that isn't. Except in other contexts you do need the cast :/

Anyway I've initialized the static data explicitly, doesn't seem worth making a new macro just for this reason.

**Pull request merged**